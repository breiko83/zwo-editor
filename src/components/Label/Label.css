.label {
  position: absolute;
  z-index: 100;
  top: -115px;
  left: 0;
  color: white;
  background-color: rgba(0, 0, 0, 0.7);
  width: 150px;
  border-radius: 5px;
  margin: 5px 0;
  text-align: center;
  font-size: 14px; 
  padding: 5px;
}

.textField {
  background-color: rgba(255, 255, 255, 0.1);
  color: white;
  border: 0;
  border-bottom: 0px solid white;
  font-size: 14px;
}

.cadence-row {
  display: flex;
}
.cadence-row > .cadenceLabel {
  font-size: 14px;
  flex: 1;
}
.cadence-row > .textField {
  min-width: 1px;
  flex: 1;
}
