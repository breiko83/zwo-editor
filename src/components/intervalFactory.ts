import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
