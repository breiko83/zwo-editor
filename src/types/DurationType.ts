export type DurationType = "time" | "distance";