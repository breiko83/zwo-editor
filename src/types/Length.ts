export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
