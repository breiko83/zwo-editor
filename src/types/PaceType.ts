export enum PaceType {
  oneMile = 0,
  fiveKm = 1,
  tenKm = 2,
  halfMarathon = 3,
  marathon = 4,
}
