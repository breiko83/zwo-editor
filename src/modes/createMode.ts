import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
