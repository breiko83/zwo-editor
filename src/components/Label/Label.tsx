import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faBolt, faClock, faRuler } from '@fortawesome/free-solid-svg-icons'
import './Label.css'
import { FreeInterval, RampInterval, SteadyInterval } from '../../types/Interval'
import { WorkoutMode } from '../../modes/WorkoutMode'
import BikeMode from '../../modes/BikeMode'
import RunMode from '../../modes/RunMode'
import * as format from '../../utils/format'

interface LabelProps {
  interval: SteadyInterval | RampInterval | FreeInterval;
  mode: WorkoutMode;
  onCadenceChange: (cadence: number) => void;
}

const Label = ({ mode, ...props }: LabelProps) => {
  return (
    <div className='label'>
      <div>
        <FontAwesomeIcon icon={faClock} fixedWidth /> {format.duration(mode.intervalDuration(props.interval))}
      </div>
      {mode instanceof BikeMode ? <BikeData mode={mode} {...props} /> : <RunData mode={mode} {...props} />}
    </div>
  );
};

function BikeData({ interval, mode, onCadenceChange }: LabelProps & { mode: BikeMode }) {
  return (
    <>
      {interval.type === "steady" &&
        <>
          <div>
            <FontAwesomeIcon icon={faBolt} fixedWidth /> {format.power(mode.power(interval.intensity))}
          </div>
          <div>
            {format.wkg(mode.wkg(interval.intensity))} &middot; {format.percentage(interval.intensity)} FTP
          </div>
        </>
      }
      {interval.type === "ramp" &&
        <>
          <div>
            <FontAwesomeIcon icon={faBolt} fixedWidth /> {format.power(mode.power(interval.startIntensity))} - {format.power(mode.power(interval.endIntensity))}
          </div>
          <div>
            {format.percentage(interval.startIntensity)} - {format.percentage(interval.endIntensity)} FTP
          </div>
        </>
      }
      <div className="cadence-row">
        <label className="cadenceLabel">Cadence</label>
        <CadenceInput cadence={interval.cadence} onCadenceChange={onCadenceChange} />
      </div>
    </>
  );
}

function RunData({ interval, mode }: LabelProps & { mode: RunMode }) {
  return (
    <>
      <div>
        <FontAwesomeIcon icon={faRuler} fixedWidth /> {format.distance(mode.intervalDistance(interval))}
      </div>
      {interval.type === "steady" &&
        <div>
          {format.percentage(interval.intensity)} {format.shortPaceName(interval.pace)}
        </div>
      }
      {interval.type === "ramp" &&
        <div>
          {format.percentage(interval.startIntensity)} to {format.percentage(interval.endIntensity)} {format.shortPaceName(interval.pace)}
        </div>
      }
    </>
  );
}

function CadenceInput({cadence, onCadenceChange}: {cadence: number, onCadenceChange: (v: number) => void}) {
  return (
    <input
      type="number"
      min="40"
      max="150"
      step="5"
      name="cadence"
      value={cadence || ''}
      onChange={(e) => {onCadenceChange(parseInt(e.target.value))}}
      onClick={(e)=> {e.stopPropagation()}}
      className="textField cadence"
    />
  );
}

export default Label