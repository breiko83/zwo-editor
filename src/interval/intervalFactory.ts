import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
