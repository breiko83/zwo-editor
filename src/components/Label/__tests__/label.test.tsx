import React from 'react';
import Label from '../Label';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import { PaceType } from '../../../types/PaceType';
import intervalFactory from '../../../interval/intervalFactory';
import createMode from '../../../modes/createMode';
import { Distance, Duration } from '../../../types/Length';

const MockReact = React;

jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon(props: any) {
    return MockReact.createElement("FontAwesomeIcon", props);
  },
}));

jest.mock('@fortawesome/free-solid-svg-icons', () => ({
  faBolt: "faBolt",
  faClock: "faClock",
  faRuler: "faRuler",
}));

describe('<Label>', () => {
  test('for cycling, renders: duration, power, w/kg, %FTP, cadence', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.steady({ length: new Duration(100), intensity: 1.25, cadence: 0 }, mode);
    const component = renderer.create(
      <Label interval={interval} mode={mode} onCadenceChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  test('for cycling ramp, renders: duration, power-range, %FTP-range, cadence', () => {
    const mode = createMode({sportType: "bike", ftp: 200, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.ramp({ length: new Duration(100), startIntensity: 0.5, endIntensity: 1.0, cadence: 0 }, mode);
    const component = renderer.create(
      <Label interval={interval} mode={mode} onCadenceChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  test('for running, renders: duration, distance, %pace, pace type', () => {
    const mode = createMode({sportType: "run", ftp: 1234, weight: 75, runningTimes: [0, 0, 1000, 0, 0], lengthType: "time"});
    const interval = intervalFactory.steady({ length: new Duration(100), intensity: 1.25, cadence: 0, pace: PaceType.tenKm }, mode);
    const component = renderer.create(
      <Label interval={interval} mode={mode} onCadenceChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  test('for running (in distance mode), renders: duration, distance, %pace, pace type', () => {
    const mode = createMode({sportType: "run", ftp: 1234, weight: 75, runningTimes: [0, 0, 1000, 0, 0], lengthType: "distance"});
    // Using the same distance as was calculated in previous test,
    // and expecting to get the same 1:40 duration calculated from it
    const interval = intervalFactory.steady({ length: new Distance(1250), intensity: 1.25, cadence: 0, pace: PaceType.tenKm }, mode);
    const component = renderer.create(
      <Label interval={interval} mode={mode} onCadenceChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });

  test('for running ramp, renders: duration, distance, %pace-range, pace type', () => {
    const mode = createMode({sportType: "run", ftp: 1234, weight: 75, runningTimes: [0, 0, 1000, 0, 0], lengthType: "time"});
    const interval = intervalFactory.ramp({ length: new Duration(100), startIntensity: 0.5, endIntensity: 1.0, cadence: 0, pace: PaceType.tenKm }, mode);
    const component = renderer.create(
      <Label interval={interval} mode={mode} onCadenceChange={() => {}} />
    );
    expect(component).toMatchSnapshot();
  });
});
