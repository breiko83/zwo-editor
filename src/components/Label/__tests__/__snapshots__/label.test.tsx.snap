// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Label> for cycling ramp, renders: duration, power-range, %FTP-range, cadence 1`] = `
<div
  className="label"
>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faClock"
    />
     
    01:40
  </div>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faBolt"
    />
     
    100 W
     - 
    200 W
  </div>
  <div>
    50%
     - 
    100%
     FTP
  </div>
  <div
    className="cadence-row"
  >
    <label
      className="cadenceLabel"
    >
      Cadence
    </label>
    <input
      className="textField cadence"
      max="150"
      min="40"
      name="cadence"
      onChange={[Function]}
      onClick={[Function]}
      step="5"
      type="number"
      value=""
    />
  </div>
</div>
`;

exports[`<Label> for cycling, renders: duration, power, w/kg, %FTP, cadence 1`] = `
<div
  className="label"
>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faClock"
    />
     
    01:40
  </div>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faBolt"
    />
     
    250 W
  </div>
  <div>
    3.3 W/kg
     Â· 
    125%
     FTP
  </div>
  <div
    className="cadence-row"
  >
    <label
      className="cadenceLabel"
    >
      Cadence
    </label>
    <input
      className="textField cadence"
      max="150"
      min="40"
      name="cadence"
      onChange={[Function]}
      onClick={[Function]}
      step="5"
      type="number"
      value=""
    />
  </div>
</div>
`;

exports[`<Label> for running (in distance mode), renders: duration, distance, %pace, pace type 1`] = `
<div
  className="label"
>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faClock"
    />
     
    01:40
  </div>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faRuler"
    />
     
    1250 m
  </div>
  <div>
    125%
     
    10K pace
  </div>
</div>
`;

exports[`<Label> for running ramp, renders: duration, distance, %pace-range, pace type 1`] = `
<div
  className="label"
>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faClock"
    />
     
    01:40
  </div>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faRuler"
    />
     
    750 m
  </div>
  <div>
    50%
     to 
    100%
     
    10K pace
  </div>
</div>
`;

exports[`<Label> for running, renders: duration, distance, %pace, pace type 1`] = `
<div
  className="label"
>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faClock"
    />
     
    01:40
  </div>
  <div>
    <FontAwesomeIcon
      fixedWidth={true}
      icon="faRuler"
    />
     
    1250 m
  </div>
  <div>
    125%
     
    10K pace
  </div>
</div>
`;
